### ISSUE TITLE
harmony.realizeChordSymbolDurations() causes misalignment when XML exporting.

### ISSUE THREAD

**Issue by DMcGilli** [cuthbertLab/music21#963]
![harmony realizeChordurations OFF-1](https://user-images.githubusercontent.com/81744582/115085543-9030ee00-9ebf-11eb-9dc1-4746ea2dd5b9.png)
![harmony RealizeChordDurations ON-1](https://user-images.githubusercontent.com/81744582/115085556-93c47500-9ebf-11eb-86d3-cbb45230efff.png)
[ChordSymbol Durations Script and MXML file.zip](https://github.com/cuthbertLab/music21/files/6327941/ChordSymbol.Durations.Script.and.MXML.file.zip)

:
Per thread from the mailing list:

I had a quick look -- this is reproducible without involving templates. If you set writeAsChord = False (which harmony.realizeChordSymbolDurations() does) then musicxml export gets gummed up if there are gaps between them and no rests to fill the gap. Please open a GitHub issue.

A workaround would be to fill with rests (use the argument provided when making the template, or just makeRests(fillGaps=True)), and then recurse over the stream and set .style.hideObjectOnPrint = True on each rest if you don't want to see them.

makeNotation() is run during musicxml export, and fixes lots of things that might gum up musicxml display, but making rests isn't one of them. Which is a neat feature for someone making a worksheet. Tangent: I've been toying with the idea of adding a makeNotation=False argument to show/write so that you can "freeze" what you have and not let musicxml export make more choices for you. But irrelevant here because makeNotation doesn't (currently) make rests.

Thanks for the report.
Jacob

**Comment by jacobtylerwalls** [comment]
**Reproducer**

```
cs1 = harmony.ChordSymbol('C7')
cs2 = harmony.ChordSymbol('F7')
m = stream.Measure()
m.insert(0, cs1)
m.insert(2, cs2)
harmony.realizeChordSymbolDurations(m)
m.write()
```

**Result**
```
    <measure number="0">
      <attributes>
        <divisions>10080</divisions>
        ...
      </attributes>
      <harmony>
        <root>
          <root-step>C</root-step>
        </root>
        <kind>dominant</kind>
        <offset sound="yes">-20160</offset>
      </harmony>
      <harmony>
        <root>
          <root-step>F</root-step>
        </root>
        <kind>dominant</kind>
      </harmony>
    </measure>
```

**Versus if durations are not realized, we create a `<forward>` tag to get us from 0.0 to 2.0**:
```
    <measure number="0">
      <attributes>
        <divisions>10080</divisions>
        ...
      </attributes>
      <harmony>
        <root>
          <root-step>C</root-step>
        </root>
        <kind>dominant</kind>
      </harmony>
      <forward>
        <duration>20160</duration>
      </forward>
      <harmony>
        <root>
          <root-step>F</root-step>
        </root>
        <kind>dominant</kind>
      </harmony>
    </measure>
```

Seems like a reasonable request to get the latter. I'm not sure if my suggested workarounds actually work.

This patch prevents the problem, but I'm not sure what it might break--would need some investigation before going further with this approach.

```diff
diff --git a/music21/musicxml/m21ToXml.py b/music21/musicxml/m21ToXml.py
index 04ee237aa..1a92fa6ae 100644
--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -2821,7 +2821,7 @@ class MeasureExporter(XMLExporterBase):
             root.append(sp)
 
         classes = obj.classes
-        if 'GeneralNote' in classes:
+        if 'GeneralNote' in classes and 'Harmony' not in classes:
             self.offsetInMeasure += obj.duration.quarterLength
```

**Comment by jacobtylerwalls** [comment]
@DMcGilli I won't be able to go much further at the moment, but if you want to collaborate in order to get this in faster, would you be willing to write a handful of tests to see if my suggested approach breaks anything?

We want .writeAsChord = False, we want a realized duration on the chord symbols, and we want to test a mixture of cases involving measures with chord symbols only, chords AND chord symbols, with and without gaps (i.e. missing rests). That's 8 permutations. Then we can add the other edge cases, like a Voice, NoChord, two at the same offset, etc.

You'd be writing those 8 or so cases basically the way I wrote "Reproducer" above. You don't have to worry about the assert statement for extracting the `<forward>` tag from the output, I can handle that -- just a comment with what you expect the result to be would be good (i.e. whether there should be a forward tag to represent a missing rest, or whether there IS a rest in your test case, etc)

If you're game, this would be VERY helpful!

**Comment by DMcGilli** [comment]
Absolutely! Let me think about what that would look like and how shall I submit them, just as .py files here? Check them in myself? 

Absolutely down to help in anyway possible!

**Comment by jacobtylerwalls** [comment]
A pull request would be great. Something like this at the bottom of m21ToXml.py:

```diff
diff --git a/music21/musicxml/m21ToXml.py b/music21/musicxml/m21ToXml.py
index 04ee237aa..d8dc3357b 100644
--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -6466,6 +6466,28 @@ class Test(unittest.TestCase):
         rest = tree.find('.//rest')
         self.assertEqual(rest.get('measure'), 'yes')
 
+    def testExportChordSymbols(self):
+        gex = GeneralObjectExporter()
+
+        # Two chord symbols, no rests
+        cs1 = harmony.ChordSymbol('C7')
+        cs2 = harmony.ChordSymbol('F7')
+        m = stream.Measure()
+        m.insert(0, cs1)
+        m.insert(2, cs2)
+        harmony.realizeChordSymbolDurations(m)
+        obj = gex.fromGeneralObject(m)
+        tree = self.getET(obj)
+        # no <offset> tags
+        self.assertEqual(len(tree.findall('.//offset')), 0)
+        # 1 <forward> tag to get from 0.0 to 2.0
+        self.assertEqual(len(tree.findall('.//forward')). 1)
+
+        # Two chord symbols, with rests
+
+        # One chord symbol midmeasure, no rests
+
+        # etc...
```

**Comment by jacobtylerwalls** [comment]
> makeNotation() is run during musicxml export, and fixes lots of things that might gum up musicxml display, but making rests isn't one of them. Which is a neat feature for someone making a worksheet.

Actually, there was an attempt to makeRests in musicxml export, but it only worked in flat streams until we merge #986, so I retested this against #986. Still gives problems, because makeRests doesn't treat a chordSymbol with a realized duration as constituting a "gap". So that gives me some confidence that my suggested approach is right (and that we don't need to test all 8 permutations, just a selection).

**Comment by DMcGilli** [comment]
Apologies for the delay but I have some initial test cases, some for importing some for creating from scratch:

https://github.com/DMcGilli/CHORDSYMBOLDURATIONS_TESTCASES

Make sure these are on the right track, and I may need guidance doing a pull request, it will be my first time.

**Comment by jacobtylerwalls** [comment]
Sure, I can give you a few pointers.

- use github.com to fork the cuthbertLab/music21 repo: this creates yourName/music21
- clone it to your computer (GItHub surely has some tutorials for this or just google clone repo)
-  see the music21 developer docs (linked on our repo homepage) if you get stuck
- create a branch `chordSymbol-duration-export` and commit your changes to it (just the music21/musicxml/m21ToXml.py file
- push to the cloud, i.e. to "origin" (google "add remote origin")
- from github.com view yourName/music21 and open a pull request against cuthbertLab/music21

Then I can look at your changes in more detail once we have that much going. Thanks again for helping out!

**Comment by jacobtylerwalls** [comment]
Ah, and one more thing -- we're not going to use visual tests; we're just going to capture the XML output in memory, assign it to a variable, and then check the tags, either by testing for substrings, or using ElementTree. I can help you with ElementTree. But the idea is to make sure that we have or don't have backup tags and rest tags, etc.. Thanks!
